searchcore/src/tests/proton/persistenceengine/resource_usage_tracker/resource_usage_tracker_test.cpp

```diff
@@ -82,4 +99,119 @@ TEST_F(ResourceUsageTrackerTest, register_guard_handles_deleted_tracker)
     _tracker.reset();
 }
 
+namespace {
+
+struct NamedAttribute
+{
+    vespalib::string subdb;
+    vespalib::string attribute;
+
+    NamedAttribute(const vespalib::string& subdb_in, const vespalib::string& attribute_in)
+        : subdb(subdb_in),
+          attribute(attribute_in)
+    {
+    }
+};
+
+NamedAttribute ready_a1("0.ready", "a1");
+NamedAttribute notready_a1("2.notready", "a1");
+NamedAttribute ready_a2("0.ready", "a2");
+
+constexpr size_t usage_limit = 1024;
+
+struct AttributeUsageStatsBuilder
+{
+    AttributeUsageStats stats;
+
+    AttributeUsageStatsBuilder()
+        : stats()
+    {
+    }
+
+    ~AttributeUsageStatsBuilder();
+
+    AttributeUsageStatsBuilder& reset() { stats = AttributeUsageStats(); return *this; }
+    AttributeUsageStatsBuilder& merge(const NamedAttribute& named_attribute, size_t used_enum_store, size_t used_multivalue);
+
+    AttributeUsageStats build() { return stats; }
+
+};
+
+AttributeUsageStatsBuilder::~AttributeUsageStatsBuilder() = default;
+
+AttributeUsageStatsBuilder&
+AttributeUsageStatsBuilder::merge(const NamedAttribute& named_attribute, size_t used_enum_store, size_t used_multivalue)
+{
+    vespalib::AddressSpace enum_store_usage(used_enum_store, 0, usage_limit);
+    vespalib::AddressSpace multivalue_usage(used_multivalue, 0, usage_limit);
+    search::AddressSpaceUsage as_usage(enum_store_usage, multivalue_usage);
+    stats.merge(as_usage, named_attribute.attribute, named_attribute.subdb);
+    return *this;
+}
+
+double rel_usage(size_t usage) noexcept {
+    return (double) usage / (double) usage_limit;
+}
+
+ResourceUsage make_resource_usage(const vespalib::string& enum_store_name, size_t used_enum_store, const vespalib::string &multivalue_name, size_t used_multivalue)
+{
+    AttributeResourceUsage enum_store_usage(rel_usage(used_enum_store), enum_store_name);
+    AttributeResourceUsage multivalue_usage(rel_usage(used_multivalue), multivalue_name);
+    return ResourceUsage(0.0, 0.0, enum_store_usage, multivalue_usage);
+}
+
+}
+
+TEST_F(ResourceUsageTrackerTest, aggregates_attribute_usage)
+{
+    notify(0.0, 0.0);
+    auto register_guard = _tracker->set_listener(*_listener);
+    auto aul1 = _tracker->make_attribute_usage_listener("doctype1");
+    auto aul2 = _tracker->make_attribute_usage_listener("doctype2");
+    AttributeUsageStatsBuilder b1;
+    AttributeUsageStatsBuilder b2;
+    b1.merge(ready_a1, 10, 20).merge(ready_a2, 5, 30);
+    b2.merge(ready_a1, 15, 15);
+    aul1->notify_attribute_usage(b1.build());
+    aul2->notify_attribute_usage(b2.build());
+    EXPECT_EQ(make_resource_usage("doctype2.0.ready.a1", 15, "doctype1.0.ready.a2", 30), _listener->get_usage());
```

Comment:
Could consider adding an _expect_resource_usage()_ on ResourceUsageTrackerTest instead to avoid using __listener_ explicit each time. The same applies for the _EXPECT_EQ((std::vector<double>{ X, Y })_ checks in previous tests.

Reply:
Line numbers for test failures will become less useful when EXPECT_EQ is moved to helper functions.

Reply:
Agreed. The solution chosen now is the best of both worlds.